### ISSUE TITLE
Large discrepancy in speaker similarity scores between ESPnet ECAPA-WavLM Joint and Microsoft WavLM

### ISSUE THREAD

**Issue by NewGamezzz** [espnet/espnet#6477]
I found a potentially suspicious behavior when using Versa to compute speaker similarity scores.

In my case, I want to evaluate a VITS TTS model trained on LibriTTS (in espnet3). In the TTS task, it is common to use the [Microsoft WavLM speaker model](https://huggingface.co/microsoft/wavlm-base-sv) to compute speaker similarity. However, Versa currently does not support general Hugging Face models and only supports models available through espnet zoo. Because of that, I chose [ECAPA-WavLM Joint](https://huggingface.co/espnet/voxcelebs12_ecapa_wavlm_joint), which seemed to be the closest available model.

Here is what I found:

Using ECAPA-WavLM Joint in Versa, I get speaker similarity scores around 0.33.
Using the Microsoft WavLM, I get scores around 0.77.

I was surprised by how different the score ranges are between the two models. I understand that the absolute values may not be directly comparable across different speaker embedding models, but the gap seems quite large. So it might be worth investigating why the score ranges are so different between these two models.

Thanks!

**Comment by sw005320** [comment]
@Jungjee and @ftshijt, did you notice it?
We may need to identify the issue

**Comment by Jungjee** [comment]
@NewGamezzz, could you please provide distributions, potentially in the form of historgrams,large enough to be statistically significant? 

On top of this, implementation of loss functions may have an impact, as well.
Also, I am wondering if you are normalizing the length of vectors when you calculate cosine similarity. 

One quick info: for out-of-domain usage, I'd recommend Wavlm without finetuning one in ESPnet. The one you used is finetuned specifically for better performance on VoxCeleb test sets.

**Comment by Jungjee** [comment]
FYI: https://huggingface.co/espnet/voxcelebs12_ecapa_frozen

**Comment by NewGamezzz** [comment]
Here, I compared the cosine similarity histograms from the Microsoft WavLM model and the ESPnet WavLM model without fine-tuning (https://huggingface.co/espnet/voxcelebs12_ecapa_frozen).

I only used 100 samples for this quick test, but the difference is already quite obvious. The histograms are in very different ranges and do not seem to overlap.

<img width="1200" height="600" alt="Image" src="https://github.com/user-attachments/assets/5d8a240f-8dca-4877-ad29-87504dc68a68" />

I also made sure to normalize the embeddings before computing cosine similarity, so I do not think the issue comes from the similarity calculation itself.

Here is the code I use for this debugging:

```python
import os

import librosa
import matplotlib.pyplot as plt
import numpy as np
import torch
import torch.nn.functional as F
from tqdm import tqdm
from transformers import AutoFeatureExtractor, AutoModelForAudioXVector
from espnet2.bin.spk_inference import Speech2Embedding

SAMPLE_RATE = 16_000               # WavLM and Espnet SPK expects 16 kHz mono.

def _load_audio(path: str) -> np.ndarray:
    """Load audio and return a 1-D float32 array at 16 kHz, mono."""
    wav, _ = librosa.load(path, sr=SAMPLE_RATE, mono=True)
    return wav.astype(np.float32)


@torch.no_grad()
def _embed(
    model,
    extractor,
    wav: np.ndarray,
    model_type: str,
    device: torch.device,
) -> torch.Tensor:
    """Run a single utterance through WavLM-Base-Plus-SV and return its
    L2-normalised x-vector (512-dim, CPU)."""
    if model_type == "microsoft":
        inputs = extractor(wav, sampling_rate=SAMPLE_RATE, return_tensors="pt")
        inputs = {k: v.to(device) for k, v in inputs.items()}
        embedding = model(**inputs).embeddings             
        return F.normalize(embedding, dim=-1).squeeze(0).cpu()
    else:
        wav = torch.from_numpy(wav).to(device) 
        embedding = model(wav)
        return F.normalize(embedding, dim=-1).squeeze(0).cpu()


def main():
    device = torch.device("cuda" if torch.cuda.is_available() else "cpu")
    print("Using device:", device)
    wavlm_extractor = AutoFeatureExtractor.from_pretrained("microsoft/wavlm-base-plus-sv")
    wavlm_model = (
        AutoModelForAudioXVector.from_pretrained("microsoft/wavlm-base-plus-sv").to(device).eval()
    )
    wavlm_sims = []

    espnet_extractor = None
    espnet_model = Speech2Embedding.from_pretrained(model_tag="espnet/voxcelebs12_ecapa_frozen")
    espnet_sims = []

    infer_folder = "egs3/libritts/tts/exp/train_vits_libritts/inference/test/wav"
    ref_metadata = "egs3/libritts/tts/exp/train_vits_libritts/inference/test/ref.scp"
    
    files_list = []
    num_samples = 100
    with open(ref_metadata) as f:
        for line in f:
            line = line.strip()
            if not line:
                continue
            fid, ref_path = line.split(maxsplit=1)
            pred_path = os.path.join(infer_folder, f"{fid}.wav")
            files_list.append((fid, pred_path, ref_path))

    for fid, pred_path, ref_path in tqdm(files_list[:num_samples], desc="Processing"):
        pred_audio = _load_audio(pred_path)
        ref_audio  = _load_audio(ref_path)
        wavlm_pred_emb   = _embed(wavlm_model, wavlm_extractor, pred_audio, "microsoft", device) # normalized embedding
        wavlm_ref_emb    = _embed(wavlm_model, wavlm_extractor, ref_audio, "microsoft", device) # normalized embedding
        wavlm_sim = float((wavlm_pred_emb * wavlm_ref_emb).sum().item()) # dot product of normalized embeddings = cosine similarity
        wavlm_sims.append(wavlm_sim)

        espnet_pred_emb = _embed(espnet_model, espnet_extractor, pred_audio, "espnet", device)
        espnet_ref_emb  = _embed(espnet_model, espnet_extractor, ref_audio, "espnet", device)
        espnet_sim = float((espnet_pred_emb * espnet_ref_emb).sum().item())
        espnet_sims.append(espnet_sim)

        print(f"{fid}: WavLM similarity = {wavlm_sim:.4f}, Espnet similarity = {espnet_sim:.4f}")

    print(f"Wavlm Mean similarity: {np.mean(wavlm_sims):.4f}")
    print(f"Espnet Mean similarity: {np.mean(espnet_sims):.4f}")

    # plot histograms comparing the two similarity distributions
    plt.figure(figsize=(12, 6))
    plt.hist(wavlm_sims, alpha=0.7, label="WavLM")
    plt.hist(espnet_sims, alpha=0.7, label="Espnet")
    plt.xlabel("Cosine Similarity")
    plt.ylabel("Frequency")
    plt.legend()
    plt.savefig("similarity_histograms.png")


if __name__ == "__main__":
    main()
```

**Comment by Jungjee** [comment]
Thanks, @NewGamezzz, for the quick turnaround. Yes, it looks significant (seems like no need to talk about whether it's statistically significant).
One more question: what does the ground truth look like? You said 100 samples, I'm assuming it would be 100 "pairs" of audio samples. Are the pairs coming from the same speaker?

**Comment by NewGamezzz** [comment]
Yes, they are 100 pairs of audio samples.

Since this is a TTS evaluation setup, each pair consists of a speaker prompt and the generated speech conditioned on that prompt. So they are intended to be same-speaker pairs. I then extract embeddings from both audio samples and compute the cosine similarity between them.

**Comment by Jungjee** [comment]
I see all positive pairs then (assuming no spoofing detection capability).
One suggestion: I'd test both models with Vox1-O or another evaluation protocol for speaker verification and see what the EER thresholds look like. 

If, for example, EER threshold of ESPnet WavLM is 0.1, the histogram makes sense.  
Another potential, try reading waveforms with int16 (e.g., soundfile.read(filename, dtype="int16")).
